lib/uefi/variable_mem.cpp

```diff
@@ -0,0 +1,198 @@
+#include <lk/list.h>
+#include <uefi/types.h>
+
+namespace {
+  struct EfiVariableInMem {
+    struct list_node node;
+    char16_t* VariableName;
+    EfiGuid VendorGuid;
+    uint32_t Attributes;
+    char* Data;
+    size_t DataLen;
+  };
+}
+
+static list_node variables_in_mem = LIST_INITIAL_VALUE(variables_in_mem);
+
+static int comp_variable_name(const char16_t *varname1,
+                              const char16_t *varname2) {
+  for (int i = 0; varname1[i] != 0 && varname2[i] != 0; i++) {
+    if (varname1[i] != varname2[i])
+      return varname1[i] - varname2[i];
+  }
+  return 0;
+}
+
+static EfiVariableInMem * search_existing_variable(const char *variable_name,
+                                                   const EfiGuid *guid) {
+  /* convert US-ASCII to UTF-16 */
+  char16_t varname[512];
+  memset(varname, 0, sizeof(varname));
+  for (size_t i = 0; i < sizeof(varname) / sizeof(varname[0]) && variable_name[i] != '\0'; i++) {
+    varname[i] = variable_name[i];
+  }
+
+  /* search for existing variable */
+  struct EfiVariableInMem *var = nullptr;
+  list_for_every_entry(&variables_in_mem, var, struct EfiVariableInMem, node) {
+    if (guid != NULL && memcmp(&var->VendorGuid, guid, sizeof(EfiGuid)) != 0) {
+      continue;
+    }
+    if (comp_variable_name(varname, var->VariableName) != 0) {
+      continue;
+    }
+    return var;
+  }
+  return nullptr;
+}
+
+EfiStatus efi_get_variable_mem(const char *variable_name,
+                               const EfiGuid *guid,
+                               uint32_t *attribute,
+                               char **data,
+                               size_t *data_size) {
+  struct EfiVariableInMem *var = nullptr;
+
+  var = search_existing_variable(variable_name, guid);
+
+  if (data)
+    *data = nullptr;
+  if (data_size)
+    *data_size = 0;
+  if (attribute)
+    *attribute = 0;
+  if (var) {
+    if (data)
+      *data = var->Data;
+    if (data_size)
+      *data_size = var->DataLen;
+    if (attribute)
+      *attribute = var->Attributes;
+    return SUCCESS;
+  }
+  return NOT_FOUND;
+}
+
+void efi_set_variable_mem(const char *variable_name,
+                          const EfiGuid *guid,
+                          uint32_t attribute,
+                          const char *data,
+                          size_t data_len) {
+  struct EfiVariableInMem *var = nullptr;
+
+  var = search_existing_variable(variable_name, guid);
+
+  /* alloc new variable if it is not existed */
+  if (!var) {
+    var = reinterpret_cast<struct EfiVariableInMem *>(malloc(sizeof(struct EfiVariableInMem)));
+    memset(var, 0, sizeof(struct EfiVariableInMem));
+    var->node = LIST_INITIAL_CLEARED_VALUE;
+    int name_len = (strlen(variable_name) + 1) * 2;
+    var->VariableName = reinterpret_cast<char16_t *>(malloc(name_len));
+    memset(var->VariableName, 0, name_len);
+    for (size_t i = 0; i < strlen(variable_name) ; i++) {
+      var->VariableName[i] = variable_name[i];
```

Comment:
caller just converted utf16 to utf8, and now we are converting back to utf16, I don't quite understand why we are doing this. Maybe just change `efi_set_variable_mem` to take utf16 directly?